### PR TITLE
[FEATURE] Script to automate proper triggering of Docs Azure pipeline

### PR DESCRIPTION
**Pull request by cdkini**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Please see ticket for context: https://superconductive.atlassian.net/browse/GDOC-175
- Script to determine what dependencies `docs/` has on `great_expectations/`.
- Implementation of this script in our Docs Azure YAML to dynamically determine when to run the pipeline.
- Tests to cover the changes made.

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Comment by cdkini** [comment]
Screen Shot 2021-07-14 at 5 52 44 PM

Output in Azure looks something like this ^^^

File: azure-pipelines-docs-integration.yml [hunk 1]
```diff
@@ -52,13 +52,27 @@ stages:
...
+
+              echo "Checking for file changes from 'great_expectations/' that impact 'docs/'..."
+              for FILE in $DEPENDENCIES; do
+                if [[ ${CHANGED_FILES[@]} =~ $FILE ]]; then
+                    echo "Found change in local dependency:" $FILE
+                    echo "##vso[task.setvariable variable=DocsDependenciesChanged;isOutput=true]true"
+                    break
+                fi
+              done
+            name: CheckDocsDependenciesChanges
```
Comment (cdkini) [inline azure-pipelines-docs-integration.yml]: Azure does not provided native support for arrays so we need to pipe the output of the script to a bash array and iterate through. 

`CHANGED_FILE` compares against `develop` to determine which files have changed. If one of those files is in the output of `trace_docs_deps.py`, the pipeline is triggered.

Should this script be moved to an external bash script? I thought it was small enough to keep in-line but happy to adjust if need be.

File: azure-pipelines-docs-integration.yml [hunk 2]
```diff
@@ -52,13 +52,27 @@ stages:
...
 
   - stage: docusaurus_tests
     pool:
       vmImage: 'ubuntu-latest'
     jobs:
     - job: test_docs
-      condition: or(eq(stageDependencies.scope_check.changes.outputs['CheckChanges.GEChanged'], true), eq(stageDependencies.scope_check.changes.outputs['CheckChanges.DocsChanged'], true), eq(variables.isMain, true))
+      condition: or(eq(stageDependencies.scope_check.changes.outputs['CheckDocsDependenciesChanges.DocsDependenciesChanged'], true), eq(stageDependencies.scope_check.changes.outputs['CheckDocsChanges.DocsChanged'], true), eq(variables.isMain, true))
```
Comment (cdkini) [inline azure-pipelines-docs-integration.yml:63]: This previously used `CheckChanges.GEChanged`, which is a condition in our primary Azure YAML. I think looking at every single change in GE is far too costly for our CI/CD.

In order for our docs pipeline to run, we must fulfill one of two conditions:
  - Something in our `docs/` changed.
  - A local dependency used within `docs/` has changed.

File: scripts/trace_docs_deps.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def _get_relative_path(line: str, doc: str) -> str:
+    pattern: str = "=(.+?)#"  # Parse just the path from the Docusaurus link
+    search: re.Match[str] = re.search(pattern, line)
+    path: str = search.group(1)
+
+    # Ensure that paths are relative to pwd
+    nesting: int = doc.count("/")
+    parts: List[str] = path.split("/")
```
Comment (cdkini) [inline scripts/trace_docs_deps.py]: This is a somewhat crude way to take the relative paths given by the Docusaurus links and adjust them to work with our `pwd`.

Thoughts on how to improve this?

File: scripts/trace_docs_deps.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    for file in files:
+        with open(file) as fh:
+            root: ast.Module = ast.parse(fh.read(), file)
+        for node in ast.walk(root):
+            if not isinstance(
+                node, ast.ImportFrom
+            ):  # ast.Import is only used for external deps
+                continue
+
+            # Only consider imports relevant to GE (note that "import great_expectations as ge" is discarded)
```
Comment (cdkini) [inline scripts/trace_docs_deps.py:73]: I deemed it appropriate to discard general `GE` imports since they would always cause the pipeline to trigger.

File: tests/scripts/test_trace_docs_deps.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def test_get_local_imports_discards_general_ge_imports(tmpdir_factory):
+    temp_dir: py.path.local = tmpdir_factory.mktemp("temp_dir")
+    temp_file: py.path.local = temp_dir.join("temp_file.py")
+    with open(temp_file, "w") as f:
+        f.write("import great_expectations\n")
+        f.write("import great_expectations as ge")
+
+    assert len(get_local_imports([str(temp_file)])) == 0
```
Comment (cdkini) [inline tests/scripts/test_trace_docs_deps.py]: I did not come up with unit tests for `get_import_paths` as the function is highly reliant on a few moving pieces. Happy to go back and update this but might want to bounce off some ideas on how best to implement.

File: scripts/trace_docs_deps.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
+"""
```
Comment (cdkini) [inline scripts/trace_docs_deps.py:1]: I'm really not a big fan of creating an entirely separate `scripts` directory for this but I didn't want to place it in `docs` since Docusaurus renders everything in there. Thoughts on where to keep this? Tests will have be adjusted accordingly.

File: tests/scripts/test_trace_docs_deps.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        "great_expectations/expectations/util.py",
+    ]
+
+
+def test_get_import_paths_references_files_in_ge_directory():
+    imports: List[str] = ["great_expectations.checkpoint.types"]
+    assert sorted(get_import_paths(imports)) == [
+        "great_expectations/checkpoint/types/__init__.py",
+        "great_expectations/checkpoint/types/checkpoint_result.py",
+    ]
```
Comment (cdkini) [inline tests/scripts/test_trace_docs_deps.py]: These tests are reliant on the contents of `great_expectations`. I feel as though mocking would be too much of a hassle for a unit test like this but let me know your thoughts.

File: tests/scripts/test_trace_docs_deps.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    with open(temp_file, "w") as f:
+        f.write(
+            "```python file=../../../../../../../../../../../../../../../a/b/c/script1.py#L1\n"
+        )
+        f.write(
+            "```python file=../../../../../../../../../../../../../../../d/e/f/script2.py#L5-10\n"
+        )
+        f.write(
+            "```yaml file=../../../../../../../../../../../../../../../g/h/i/script3.py#L100-200"
+        )
```
Comment (cdkini) [inline tests/scripts/test_trace_docs_deps.py]: These file references are incredibly nested due to the fact that `tmpdir_factory` creates quite a convoluted path with `mktemp()`.

File: scripts/trace_docs_deps.py (hunk 6 above)
Comment (allensallinger) [reply scripts/trace_docs_deps.py:1]: A scripts directory is not necessarily a bad thing for CI/CD, it can help make the pipeline a bit more readable if it calls scripts instead of having all the bash written directly in the pipeline. It does add an additional hop while troubleshooting but it allows you to trace through the pipeline a bit faster as well.

Comment (cdkini) [reply scripts/trace_docs_deps.py:1]: Hmm that's a good point! I guess I was reluctant to make a directory in the root of the project but doubt it's a huge deal.

Comment (allensallinger) [reply scripts/trace_docs_deps.py:1]: Also if the azure pipelines don't have to be defined at the top directory of the repo they could be in their own directory. Then the scripts and all things related to the deployment process could live there as well, possibly under a scripts or bin subfolder.

File: azure-pipelines-docs-integration.yml [hunk 9]
```diff
@@ -52,13 +52,28 @@ stages:
...
+              printf '%s\n' "[DEPENDENCIES]" "${DEPENDENCIES[@]}" ""
+
+              echo "Checking for file changes from 'great_expectations/' that impact 'docs/'..."
+              for FILE in $DEPENDENCIES; do
+                if [[ ${CHANGED_FILES[@]} =~ $FILE ]]; then
+                    echo "Found change in local dependency:" $FILE
+                    echo "##vso[task.setvariable variable=DocsDependenciesChanged;isOutput=true]true"
+                fi
+              done
+            name: CheckDocsDependenciesChanges
```
Comment (cdkini) [inline azure-pipelines-docs-integration.yml]: May move this all to a separate script and store in `scripts/` dir if that's the direction we're going.

File: scripts/trace_docs_deps.py (hunk 6 above)
Comment (cdkini) [reply scripts/trace_docs_deps.py:1]: Ah I really like the idea of decluttering the top directory; perhaps not fully in scope for our current sprints but I'll make note of it in a ticket. Thanks!

File: scripts/trace_docs_deps.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+def _get_relative_path(line: str, doc: str) -> str:
+    pattern: str = "=(.+?)#"  # Parse just the path from the Docusaurus link
+    search: re.Match[str] = re.search(pattern, line)
+    path: str = search.group(1)
+
+    # Ensure that paths are relative to pwd
+    depth: int = doc.count("/")
+    parts: List[str] = path.split("/")
+
+    return "/".join(part for part in parts[depth:])
```
Comment (cdkini) [inline scripts/trace_docs_deps.py]: @alexsherstinsky we must have dealt with relative pathing somewhere in GE right? There's a more elegant solution out there that I'm missing.

File: tests/scripts/test_trace_docs_deps.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+#         "great_expectations/expectations/util.py",
+#     ]
+
+
+# def test_get_import_paths_references_files_in_ge_directory():
+#     imports: List[str] = ["great_expectations.checkpoint.types"]
+#     assert sorted(get_import_paths(imports)) == [
+#         "great_expectations/checkpoint/types/__init__.py",
+#         "great_expectations/checkpoint/types/checkpoint_result.py",
+#     ]
```
Comment (cdkini) [inline tests/scripts/test_trace_docs_deps.py]: Work locally but break under Azure. I think all these tests can/should be tightened.

File: azure-pipelines-docs-integration.yml [hunk 12]
```diff
@@ -52,13 +52,28 @@ stages:
...
 
+          - bash: |
+              CHANGED_FILES=$(git diff HEAD origin/develop --name-only)
+              printf '%s\n' "[CHANGED FILES]" "${CHANGED_FILES[@]}" ""
+
+              DEPENDENCIES=$(python scripts/trace_docs_deps.py docs)
+              printf '%s\n' "[DEPENDENCIES]" "${DEPENDENCIES[@]}" ""
+
+              echo "Checking for file changes from 'great_expectations/' that impact 'docs/'..."
+              for FILE in $DEPENDENCIES; do
```
Comment (alexsherstinsky) [inline azure-pipelines-docs-integration.yml]: Just as a stylistic thought, I use `${VARIABLE}` -- so would be `for FILE in ${DEPENDENCIES}` -- personal prefernece.

File: azure-pipelines-docs-integration.yml (hunk 2 above)
Comment (alexsherstinsky) [reply azure-pipelines-docs-integration.yml:63]: Brilliant!

File: tests/scripts/test_trace_docs_deps.py (hunk 8 above)
Comment (alexsherstinsky) [reply tests/scripts/test_trace_docs_deps.py]: @cdkini Can there be a way to use the `file_relative_path()` type utility we have in order to avoid keeping track of the directory levels?

File: tests/scripts/test_trace_docs_deps.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+def test_find_docusaurus_refs_does_not_duplicate(tmpdir_factory):
+    docs_dir: py.path.local = tmpdir_factory.mktemp("docs")
+    temp_file: py.path.local = docs_dir.join("temp_file.md")
+    with open(temp_file, "w") as f:
+        f.write(
+            "```python file=../../../../../../../../../../../../../../../a/b/c/script_a.py#L1\n"
+        )
+        f.write(
+            "```python file=../../../../../../../../../../../../../../../a/b/c/script_a.py#L5-10"
+        )
```
Comment (alexsherstinsky) [inline tests/scripts/test_trace_docs_deps.py]: @cdkini Can there be a way to use the file_relative_path() type utility we have in order to avoid keeping track of the directory levels?

File: tests/scripts/test_trace_docs_deps.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    temp_file: py.path.local = temp_dir.join("temp_file.py")
+    with open(temp_file, "w") as f:
+        f.write("import great_expectations\n")
+        f.write("import great_expectations as ge")
+
+    files: List[str] = [str(temp_file)]
+    assert len(get_local_imports(files)) == 0
+
+
+# def test_get_import_paths_references_ge_files():
```
Comment (alexsherstinsky) [inline tests/scripts/test_trace_docs_deps.py]: If we do not need this commented code, should we just delete it?

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
I really like this -- just left some comments.

File: azure-pipelines-docs-integration.yml (hunk 12 above)
Comment (cdkini) [reply azure-pipelines-docs-integration.yml]: I like this! I've made a corresponding change.

File: tests/scripts/test_trace_docs_deps.py (hunk 8 above)
Comment (cdkini) [reply tests/scripts/test_trace_docs_deps.py]: Yup this is a great idea! I just need to ensure that tests still work as intended.

File: scripts/check_for_docs_deps_changes [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+DEPENDENCIES=$(python scripts/trace_docs_deps.py docs)
+printf '%s\n' "[DEPENDENCIES]" "${DEPENDENCIES[@]}" ""
+
+echo "Checking for file changes from 'great_expectations/' that impact 'docs/'..."
+for FILE in ${DEPENDENCIES}; do
+    if [[ ${CHANGED_FILES[@]} =~ $FILE ]]; then
+        echo "Found change in local dependency:" $FILE
+        echo "##vso[task.setvariable variable=DocsDependenciesChanged;isOutput=true]true"
+    fi
+done
```
Comment (cdkini) [inline scripts/check_for_docs_deps_changes:15]: @alexsherstinsky I've added the styling you requested after moving this script to a separate file.

I've got a question about this: `CHANGED_FILES=$(git diff HEAD origin/develop --name-only)`. Should this be `$(git diff HEAD HEAD~ --name-only)`? 

If we're writing an extensive PR that has a single change in one of these dependencies, every single commit in that PR will trigger the pipeline (due to `diff` evaluating changes between branches and not just changes from the most recent commit). Is there a mechanism in Azure that kills pipelines early if we stack up too many?

File: scripts/trace_docs_deps.py (hunk 10 above)
Comment (alexsherstinsky) [reply scripts/trace_docs_deps.py]: We have utilities for a number of such operations.  If you look in tests/conftest.py or some other tests (like checkpoints) and even Rule-Based Profiler, you might find the use of those...  There are also nice globbed directory listing functions in the DataConnector utilities -- please feel free to check it out; happy to help.  Thanks!

File: tests/scripts/test_trace_docs_deps.py (hunk 11 above)
Comment (alexsherstinsky) [reply tests/scripts/test_trace_docs_deps.py]: I understand how this can be frustrating -- I have been bitten by it more times than I can remember.  One possibility is that Azure virtual machines are installed with not exactly the same variant of the OS as our local development machines.  If items come out of order, you might wish to enforce your own sorting order as long as there is no loss of generality.  I had to do this for some DataConector functionality (under one of the Path Data Connector classes).

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@cdkini I left some more comments -- hope they are useful.

File: scripts/trace_docs_deps.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    2. Goes to each linked file and uses AST to parse imports used there
+    3. Filters for only relative imports and determines the paths to those files
+
+The resulting output list is all of the dependencies `docs/` has on the primary `great_expectations/` directory.
+If a change is identified in any of these files during the pipeline runtime, we know that a docs dependency has possibly
+been impacted and the pipeline should run to ensure adequate test coverage.
+
+TODO: ADD WARNING ABOUT BAD PATTERNS
+
+"""
```
Comment (cdkini) [inline scripts/trace_docs_deps.py]: Per discussion with @alexsherstinsky , we should add a note about possible edge cases and what the repercussions are.

File: scripts/check_for_docs_deps_changes [hunk 17]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+CHANGED_FILES=$(git diff HEAD origin/develop --name-only)
+printf '%s\n' "[CHANGED FILES]" "${CHANGED_FILES[@]}" ""
+
+pip install -e . > /dev/null 2>&1 # Necessary to get GE helper methods working for `trace_docs_deps.py`
+
```
Comment (cdkini) [inline scripts/check_for_docs_deps_changes]: @alexsherstinsky we actually need to install GE if we want to use the `file_relative_path()` helper method in our dependencies script.

I don't think this is worth it as the remainder of script works as intended without any external dependencies.

File: scripts/trace_docs_deps.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+Please note that this solution has some inherit disclaimers one should be familiar with. As there is no actual running
+of source code (but rather parsing of the source file's AST), it is hard to exactly pinpoint the location of a given
+import. Utilizing the `__file__` attr would provide the exact path of an import but without that, we need to assume that
+an ast.Import node corresponds to an accurate path in our codebase (i.e. `great_expectations.core.batch` results in
+great_expectations/core/batch.py).
+
+As long as imports correspond to the relative path of modules, this method will continue to work as intended. Deviations
+here would result in certain imports not being picked up by the script, reducing the accuracy and frequency of the
+Azure Docs Integration pipeline's triggers. As GE has maintained this consistency throughout its development, this
+script and the methods used herein are deemed to be appropriate.
```
Comment (cdkini) [inline scripts/trace_docs_deps.py]: @alexsherstinsky added some color around possible pitfalls. Let me know if that's adequate or if I can better clarify anything.

File: scripts/trace_docs_deps.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+    linked_files: Set[str] = set()
+    pattern: str = (
+        r"\`\`\`[a-zA-Z]+ file"  # Format of internal links used by Docusaurus
+    )
+
+    for doc in glob.glob(f"{dir}/**/*.md", recursive=True):
+        for line in open(doc):
+            if re.search(pattern, line):
+                file: str = _parse_file_from_docusaurus_link(line)
+                path: str = file_relative_path(doc, file)
```
Comment (cdkini) [inline scripts/trace_docs_deps.py]: I think this should be `os.path.join(os.path.dirname(doc), file)` (what `file_relative_path` is actually doing) so we can remove GE as a dependency from this script.

**Review by cdkini — COMMENTED** [review]
Made changes per our last discussion @alexsherstinsky please take a quick look whenever you have a minute.

I've left a comment or two but we should hopefully be good to have this merged in shortly!

**Review by alexsherstinsky — APPROVED** [review]
LGTM
